### PR TITLE
Move hash computation so that it is recomputed on retry, and now-inva…

### PR DESCRIPTION
**Pull request by alecw**
…lid checkpoint is not loaded.

If number of tries is exhausted, and ELBO tests are still failing, allow to complete anyway (using checkpoint) so that outputs are produced, but exit(1).

### CODE REVIEW

File: cellbender/remove_background/run.py [hunk 1]
```diff
@@ -59,6 +61,22 @@ def run_remove_background(args: argparse.Namespace) -> Posterior:
...
+    hashcode = create_workflow_hashcode(
+        module_path=os.path.dirname(cellbender.__file__),
+        args_to_remove=(['output_file', 'fpr', 'input_checkpoint_tarball', 'debug',
+                         'posterior_batch_size', 'checkpoint_min', 'truth_file',
+                         'posterior_regularization', 'cdf_threshold_q', 'prq_alpha',
+                         'estimator', 'use_multiprocessing_estimation', 'cpu_threads',
+                         # The following settings do not affect the results, and can change when retrying,
+                         # so remove them.
+                         'epoch_elbo_fail_fraction', 'final_elbo_fail_fraction',
+                         'num_failed_attempts', 'checkpoint_filename']
```
Comment (sjfleming) [inline cellbender/remove_background/run.py:74]: Thank you!  These had been overlooked by me

File: cellbender/remove_background/run.py [hunk 2]
```diff
@@ -771,7 +789,12 @@ def run_inference(dataset_obj: SingleCellRNACountsDataset,
                 sys.exit(0)
             else:
                 logger.info(f'No more attempts are specified by --num-training-tries. '
-                            f'Therefore the workflow will abort here.')
+                            f'Therefore the workflow will run once more without ELBO restrictions.')
```
Comment (sjfleming) [inline cellbender/remove_background/run.py:792]: Do you want to re-run it?  Does this re-run "cache", i.e. use the checkpoint?  Or does it actually retrain the whole thing?

File: cellbender/remove_background/run.py [hunk 3]
```diff
@@ -771,7 +789,12 @@ def run_inference(dataset_obj: SingleCellRNACountsDataset,
...
+                args.epoch_elbo_fail_fraction = None
+                args.final_elbo_fail_fraction = None
+                run_remove_background(args)  # start from scratch
+                # non-zero exit status in order to draw user's attention to the fact that ELBO tests
+                # were never satisfied.
                 sys.exit(1)
```
Comment (sjfleming) [inline cellbender/remove_background/run.py:798]: I can see the value of still wanting to exit 1 so that the run will be flagged as a failure

**Review by sjfleming — COMMENTED** [review]
Looks good, thank you!  One question about the final re-run strategy

File: cellbender/remove_background/run.py (hunk 2 above)
Comment (alecw) [reply cellbender/remove_background/run.py:792]: Yes, after the changes to the elements that are included in the hash computation, this uses the most recent checkpoint.

Comment (sjfleming) [reply cellbender/remove_background/run.py:792]: Awesome, sounds perfect

**Review by sjfleming — APPROVED** [review]

**Comment by sjfleming** [comment]
This looks good to me!  Thanks for contributing it.

I think you're the main person using this functionality, so if the behavior gives you what you want in your testing, I'm happy to merge this.  Unfortunately I don't have any unit tests written yet for this kind of retry functionality... some day!

**Comment by alecw** [comment]
Hi @sjfleming ,

Yes, it appears to be working fine, so please go ahead and merge.

Thanks, Alec

**Review by mbabadi — APPROVED** [review]

**Pull request merged**